activation/wire/wire_v2_test.go

```diff
@@ -0,0 +1,233 @@
+				},
+				{
+					Membership: MerkleProofV2{
+						Nodes:       make([]types.Hash32, 32),
+						LeafIndices: make([]uint64, 256),
+					},
+					Challenge: types.RandomHash(),
+					Posts:     make([]SubPostV2, 256), // actually the sum of all posts in `NiPosts` should be 256
+				},
+			},
+		}
+		for i := range atx.NiPosts[0].Posts {
+			atx.NiPosts[0].Posts[i].Post = PostV1{
+				Nonce:   0,
+				Indices: make([]byte, 800),
+				Pow:     0,
+			}
+		}
+		for i := range atx.NiPosts[1].Posts {
+			atx.NiPosts[1].Posts[i].Post = PostV1{
+				Nonce:   0,
+				Indices: make([]byte, 800),
+				Pow:     0,
+			}
+		}
+		atx.VRFNonce = new(uint64)
+		atx.MarriageATX = new(types.ATXID)
+		b.StartTimer()
+		atx.ID()
+	}
+}
+
+func Test_NoATXv2IDCollisions(t *testing.T) {
+	f := fuzz.New()
+
+	atxIDs := make([]types.ATXID, 0, 1000)
+	for range 1000 {
+		atx := &ActivationTxV2{}
+		f.Fuzz(atx)
+		id := atx.ID()
+		require.NotContains(t, atxIDs, id, "ATX ID collision")
+		atxIDs = append(atxIDs, id)
+	}
+}
+
+const PublishEpochIndex = 0
+
+func Test_GenerateDoublePublishProof(t *testing.T) {
+	atx := &ActivationTxV2{
+		PublishEpoch:   10,
+		PositioningATX: types.RandomATXID(),
+		PreviousATXs:   make([]types.ATXID, 1),
+		NiPosts: []NiPostsV2{
+			{
+				Membership: MerkleProofV2{
+					Nodes:       make([]types.Hash32, 32),
+					LeafIndices: make([]uint64, 256),
+				},
+				Challenge: types.RandomHash(),
+				Posts: []SubPostV2{
+					{
+						MarriageIndex: rand.Uint32N(256),
+						PrevATXIndex:  0,
+						Post: PostV1{
+							Nonce:   0,
+							Indices: make([]byte, 800),
+							Pow:     0,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	proof, err := generatePublishEpochProof(atx)
+	require.NoError(t, err)
+	require.NotNil(t, proof)
+
+	// a malfeasance proof for double publish will contain
+	// - the value of the PublishEpoch (here 10) - 4 bytes
+	// - the two ATX IDs - 32 bytes each
+	// - the two signatures (atx.Signature + atx.NodeID) - 64 bytes each
+	// - two merkle proofs - one per ATX - that is 128 bytes each (4 * 32)
+	// total: 452 bytes instead of two full ATXs (> 20 kB each in the worst case)
+
+	publishEpochLeaf := (*[4]byte)(unsafe.Pointer(&atx.PublishEpoch))[:]
+	ok, err := merkle.ValidatePartialTree(
+		[]uint64{PublishEpochIndex},
+		[][]byte{publishEpochLeaf},
+		proof,
+		atx.ID().Bytes(),
+		atxTreeHash,
+	)
+	require.NoError(t, err)
+	require.True(t, ok)
+
+	// different PublishEpoch doesn't validate
+	publishEpochLeaf = []byte{0xFF, 0x00, 0x00, 0x00}
+	ok, err = merkle.ValidatePartialTree(
+		[]uint64{PublishEpochIndex},
+		[][]byte{publishEpochLeaf},
+		proof,
+		atx.ID().Bytes(),
+		atxTreeHash,
+	)
+	require.NoError(t, err)
+	require.False(t, ok)
+}
+
+func generatePublishEpochProof(atx *ActivationTxV2) ([][]byte, error) {
+	tree, err := merkle.NewTreeBuilder().
+		WithLeavesToProve(map[uint64]bool{PublishEpochIndex: true}).
+		WithHashFunc(atxTreeHash).
+		Build()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
It would be useful to have a helper function that prepares the "ATX ID tree" that takes optional "indices to prove" argument. It would be reused between the malfeasance proof functions and the `atx.ID()`. WDYT? Like:

```go
func (atx *ActivationTxV2) MerkleTree(indices ...uint64) (*merkle.Tree, err) {
    idxMap := ... // turn to set
    tree, err := merkle.NewTreeBuilder().
		WithLeavesToProve(idxMap).
		WithHashFunc(atxTreeHash).
		Build()
    // ...
    return tree, nil
}

func (atx *ActivationTxV2) ID() types.ATXID {
    tree, err := atx.MerkleTree()
    if err != nil { panic("ops")}
    return types.ATXID(tree.Root())
}

func generatePublishEpochProof(atx *ActivationTxV2) ([][]byte, error) {
    tree, err := atx.MerkleTree(PublishEpochIndex)
    if err != nil {
        return nil, err
    }
    return tree.Proof(), nil
}
```

Reply:
I had a similar but different approach in mind, where I `markelTree` is a private method that I pass the `*merkle.Tree` object to so the caller can create the tree with the settings they need.